File: packing_interface/src/algorithm_templates/python/root/sleator.py
import packing_lib
from typing import List, Tuple


class Packing:
    def solve(self, bin_width: int, rectangles: List[Tuple[int, int, int]]) -> List[Tuple[float, float, int, int]]:
        items = packing_lib.expand_items(rectangles)
        wide_items = []
        remaining_items = []

        for item in items:
            width = item["width"]
            if width > bin_width:
                raise ValueError(f"Rectangle width {width} exceeds bin width {bin_width}.")
            if width > bin_width / 2:
                wide_items.append(item)
            else:
                remaining_items.append(item)

        placements = []
        current_y = 0.0

        for item in wide_items:
            placements.append((0.0, current_y, item["width"], item["height"]))
            current_y += item["height"]

        remaining_items = packing_lib.sort_by_height(remaining_items)
        half_width = bin_width / 2.0

        left_baseline = current_y
        right_baseline = current_y
        level_buffer = []
        level_used_width = 0.0
        level_height = 0.0

        def flush_full_width_level() -> None:
            nonlocal current_y, left_baseline, right_baseline, level_buffer, level_used_width, level_height
            if not level_buffer:
                return
            x = 0.0
            for width, height in level_buffer:
                placements.append((x, current_y, width, height))
                x += width
            current_y += level_height
            left_baseline = current_y
            right_baseline = current_y
            level_buffer = []
            level_used_width = 0.0
            level_height = 0.0


        idx = 0
        while idx < len(remaining_items):
            item = remaining_items[idx]
            width = item["width"]
            height = item["height"]

            if level_used_width + width <= bin_width:
                level_buffer.append((width, height))
                level_used_width += width
                level_height = max(level_height, height)
                idx += 1
                continue

            break

        flush_full_width_level()

        while idx < len(remaining_items):
            item = remaining_items[idx]
            width = item["width"]
            height = item["height"]

            place_left = left_baseline <= right_baseline
            baseline = left_baseline if place_left else right_baseline
            x_start = 0.0 if place_left else half_width
            used_width = 0.0
            level_items = []
            level_height = 0.0

            while idx < len(remaining_items):
                item = remaining_items[idx]
                width = item["width"]
                height = item["height"]
                if width > half_width:
                    break
                if used_width + width > half_width:
                    break
                level_items.append((width, height))
                used_width += width
                level_height = max(level_height, height)
                idx += 1

            if not level_items:
                fallback_y = max(left_baseline, right_baseline)
                fallback_placements = packing_lib.place_items_ffdh(
                    remaining_items[idx:],
                    bin_width,
                    start_y=fallback_y,
                )
                placements.extend(fallback_placements)
                break

            x = x_start
            for width, height in level_items:
                placements.append((x, baseline, width, height))
                x += width

            new_baseline = baseline + level_height
            if place_left:
                left_baseline = new_baseline
            else:
                right_baseline = new_baseline

        return packing_lib.output_from_placements(bin_width, placements)

File: packing_interface/src/algorithm_templates/cpp/root/sleator.cpp
#include "packing_lib.h"
using namespace packing;

class Packing {
public:
    std::vector<std::tuple<double, double, int, int>> solve(
        int binWidth,
        const std::vector<std::tuple<int, int, int>>& rectangles
    ) {
        auto items = expand_items(rectangles);
        std::vector<Rect> wideItems;
        std::vector<Rect> remainingItems;

        for (const auto& item : items) {
            if (item.width > binWidth) {
                throw std::runtime_error("Rectangle width exceeds bin width.");
            }
            if (item.width > binWidth / 2.0) {
                wideItems.push_back(item);
            } else {
                remainingItems.push_back(item);
            }
        }

        std::vector<std::tuple<double, double, int, int>> placements;
        double currentY = 0.0;

        for (const auto& item : wideItems) {
            placements.push_back({0.0, currentY, item.width, item.height});
            currentY += item.height;
        }

        remainingItems = sort_by_height(std::move(remainingItems));
        double halfWidth = static_cast<double>(binWidth) / 2.0;
        double leftBaseline = currentY;
        double rightBaseline = currentY;

        std::vector<std::pair<double, double>> levelBuffer;
        double levelUsedWidth = 0.0;
        double levelHeight = 0.0;

        auto appendFfdhFallback = [&](const std::vector<Rect>& fallbackItems, double startY) {
            struct Level {
                double y;
                int height;
                int usedWidth;
            };

            std::vector<Level> levels;
            double currentFallbackY = startY;
            for (const auto& fallbackItem : fallbackItems) {
                bool placed = false;
                for (auto& level : levels) {
                    if (level.usedWidth + fallbackItem.width > binWidth) {
                        continue;
                    }
                    placements.push_back({
                        static_cast<double>(level.usedWidth),
                        level.y,
                        fallbackItem.width,
                        fallbackItem.height,
                    });
                    level.usedWidth += fallbackItem.width;
                    placed = true;
                    break;
                }

                if (placed) {
                    continue;
                }

                levels.push_back({currentFallbackY, fallbackItem.height, fallbackItem.width});
                placements.push_back({0.0, currentFallbackY, fallbackItem.width, fallbackItem.height});
                currentFallbackY += fallbackItem.height;
            }
        };

        auto flushFullWidthLevel = [&]() {
            if (levelBuffer.empty()) {
                return;
            }
            double x = 0.0;
            for (const auto& [width, height] : levelBuffer) {
                placements.push_back({x, currentY, static_cast<int>(width), static_cast<int>(height)});
                x += width;
            }
            currentY += levelHeight;
            leftBaseline = currentY;
            rightBaseline = currentY;
            levelBuffer.clear();
            levelUsedWidth = 0.0;
            levelHeight = 0.0;
        };

        std::size_t idx = 0;
        while (idx < remainingItems.size()) {
            const auto& item = remainingItems[idx];
            if (levelUsedWidth + item.width > binWidth) {
                break;
            }
            levelBuffer.push_back({static_cast<double>(item.width), static_cast<double>(item.height)});
            levelUsedWidth += item.width;
            levelHeight = std::max(levelHeight, static_cast<double>(item.height));
            ++idx;
        }

        flushFullWidthLevel();

        while (idx < remainingItems.size()) {
            bool placeLeft = leftBaseline <= rightBaseline;
            double baseline = placeLeft ? leftBaseline : rightBaseline;
            double xStart = placeLeft ? 0.0 : halfWidth;
            double usedWidth = 0.0;
            double halfLevelHeight = 0.0;
            std::vector<std::pair<double, double>> levelItems;

            while (idx < remainingItems.size()) {
                const auto& item = remainingItems[idx];
                if (item.width > halfWidth || usedWidth + item.width > halfWidth) {
                    break;
                }
                levelItems.push_back({static_cast<double>(item.width), static_cast<double>(item.height)});
                usedWidth += item.width;
                halfLevelHeight = std::max(halfLevelHeight, static_cast<double>(item.height));
                ++idx;
            }

            if (levelItems.empty()) {
                auto fallbackItems = std::vector<Rect>(remainingItems.begin() + static_cast<long>(idx), remainingItems.end());
                double fallbackY = std::max(leftBaseline, rightBaseline);
                appendFfdhFallback(fallbackItems, fallbackY);
                break;
            }

            double x = xStart;
            for (const auto& [width, height] : levelItems) {
                placements.push_back({x, baseline, static_cast<int>(width), static_cast<int>(height)});
                x += width;
            }

            if (placeLeft) {
                leftBaseline = baseline + halfLevelHeight;
            } else {
                rightBaseline = baseline + halfLevelHeight;
            }
        }

        return placements;
    }
};
